src/app/_hooks/useCollapseRange.ts

```diff
@@ -0,0 +1,85 @@
+import { useMemo } from 'react'
+
+// Inspiration: https://atlassian.design/components/pagination
+
+const DISTANCE_FROM_START = 3
+const DISTANCE_FROM_END = 2
+const ELLIPSIS = '...'
+
+export function useCollapseRange(
+  pageCount: number,
+  currentPage: number,
+  range: number = 7,
+): Array<number | typeof ELLIPSIS> {
+  const pageNumbers = useMemo(() => {
+    return Array.from({ length: pageCount }, (_, i) => i + 1)
+  }, [pageCount])
+
+  // Boolean
+  const needEllipsis = pageNumbers.length > range
+  const needsStartEllipsis = currentPage - DISTANCE_FROM_START > 0
+  const needsEndEllipsis = currentPage + DISTANCE_FROM_END < pageNumbers.length
+
+  // page numbers
+  const firstPageNumber = pageNumbers[0]
+  const lastPageNumber = pageNumbers[pageNumbers.length - 1]
+
+  // Indexes of page numbers
+  const firstPageNumberIndex = pageNumbers.indexOf(firstPageNumber)
+  const currentPageIndex = pageNumbers.indexOf(currentPage)
+
+  // Early returns
+  if (range <= 0) throw new Error('range must be greater than 0')
+  if (!needEllipsis) return pageNumbers
+  if (range === 1) return [currentPage]
+
+  // If range is less than 4, we don't show any ellipsis
+  if (range <= 4) {
+    const firstOrBeforeCurrentIndex = Math.max(
+      firstPageNumberIndex,
+      currentPageIndex - Math.floor(range / 2),
+    )
+
+    const lastOrAfterCurrentIndex = Math.min(
+      pageNumbers.length,
+      currentPageIndex + Math.ceil(range / 2),
+    )
+
+    return pageNumbers.slice(
+      Math.min(firstOrBeforeCurrentIndex, pageNumbers.length - range),
+      Math.max(lastOrAfterCurrentIndex, range),
+    )
+  }
+
+  // Ellipsis at the start only
+  if (needsStartEllipsis && !needsEndEllipsis) {
```

Comment:
If we're writing comments, it's a good indication that the code could be clearer.

This condition, `needsStartEllipsis && !needsEndEllipsis`, can be defined in a variable.

Reply:
> If we're writing comments, it's a good indication that the code could be clearer.

I see what you mean, but for long/complex code, I think comments add a visual hierarchy to the file so that it's more accessible for the reader, in the same way headings structure paragraphs.

Thoughs?

Reply:
I don't agree. If the code is long and complex, then it should be broken up in shorter and simpler code. Abstract bits into functions, etc. I recommend reading chapter 4 from Clean Code by Robert C. Martin. (I recommend the entire book - until about chapter 8 - After that, it's very Java specific if I remember correctly.)

>So when you find yourself in a position where you need to write a comment, think it through and see whether there isn’t some way to turn the tables and express yourself in code. Every time you express yourself in code, you should pat yourself on the back. Every time you write a comment, you should grimace and feel the failure of your ability of expression.